hawkular-inventory-impl-tinkerpop-parent/hawkular-inventory-impl-tinkerpop/src/main/java/org/hawkular/inventory/impl/tinkerpop/FilterVisitor.java

```diff
@@ -85,7 +90,7 @@ public void visit(GraphTraversal<?, ?> query, Related related, QueryTranslationS
                 if (null != related.getRelationshipName()) {
                     state.setInEdges(true);
                     state.setComingFrom(Direction.IN);
-                    query.inE(related.getRelationshipName());
+                    query.inE(related.getRelationshipName()).discriminate(discriminator);
```

Comment:
I think it could be done using some decorator strategy of the graph. I didn't actually know about that possibility until a couple of days ago (I thought that Tinkerpop uses graph strategies solely for query optimization but apparently it can also use them for adding "application logic" to the traversals).

That said I am not sure if such global thing would work, because I have a feeling we do a little bit of special handling here and there that would does not follow otherwise uniform application of those checks. But it is a great point and we should definitely investigate the possibility.